src/Build/Evaluation/Conditionals/Scanner.cs

```diff
@@ -333,24 +334,40 @@ private static bool ScanForPropertyExpressionEnd(string expression, int index, o
                         if (character == '(')
                         {
                             nestLevel++;
-                            whitespaceCheck = true;
                         }
                         else if (character == ')')
                         {
                             nestLevel--;
-                            whitespaceCheck = false;
                         }
-                        else if (whitespaceCheck && char.IsWhiteSpace(character) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                        else if (char.IsWhiteSpace(character))
                         {
+                            whitespaceFound = true;
                             indexResult = index;
-                            return false;
+                        }
+                        else if (!XmlUtilities.IsValidSubsequentElementNameCharacter(character))
+                        {
+                            nonIdentifierCharacterFound = true;
+                        }
+
+                        if (character == '$')
+                        {
+                            if (!ScanForPropertyExpressionEnd(expression, index + 1, out index))
```

Comment:
Adding recursion makes me nervous but I suspect it's cheaper than keeping track of a list of bools for each `nestLevel`.